### DISCUSSION TITLE
Loading domain and problem files

### DISCUSSION THREAD

**Discussion by ko56** [shop-planner/shop3#61] [Q&A] [answered]
I'm a prospective user of SHOP3, but at the moment I'm stuck at trying to execute an example.
(My knowledge of Lisp is very rudimentary.)  So I've created the following script:

#!/usr/local/bin/sbcl --script
(require "asdf")
(asdf:load-system "shop3")

(use-package :shop3)
(use-package :common-lisp)

; Load a domain and a problem defn.:
(load "~/build/shop3/shop3/examples/rovers/strips/domain.lisp")
(load "~/build/shop3/shop3/examples/rovers/strips/p01.lisp")
(find-plans 'ROVERPROB1234 :verbose 0 :domain 'ROVER)

The result is 

SHOP3 version 3.4.0
Copyright (C) 2004-2019 SIFT, LLC.

Original SHOP2 code Copyright (C) 2002  University of Maryland.

This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
KIND, either express or implied.  This software is distributed under an
MPL/GPL/LGPL triple license.  For details, see the software source file.

Defining domain ROVER...

Defining problem ROVERPROB1234 ...
Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                    {1001580143}>:
  No such problem: ROVERPROB1234

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {1001580143}>

So I must be misunderstanding something basic.
Thanks for your help.

**Comment by rpgoldman** [comment, marked as answer]
@ko56 You have run afoul of the Common Lisp package (namespace) feature.

Your script is (implicitly) in the `common-lisp-user` package.  Looking at the head of the file "p01.lisp" in `examples/rovers/strips/` we see:

    (IN-PACKAGE :SHOP2-ROVERS)

That is why you cannot see the problem: it is in the `shop2-rovers` package.

Here's a revised script.  Try this using:

    SHOP_DIR = <YOUR SHOP DIRECTORY> sbcl --script <SCRIPTNAME>

(this assumes that you have checked out SHOP from Git, rather than using Quicklisp)

```
#!/usr/local/bin/sbcl --script
(require "asdf")

(asdf:initialize-source-registry
 `(:source-registry (:tree ,(concatenate 'string (uiop:getenv "SHOP_DIR") "/shop3"))
                     (:tree ,(concatenate 'string (uiop:getenv "SHOP_DIR") "/jenkins"))
                    :inherit-configuration))

(asdf:load-system "shop3")

(asdf:load-system "shop3/rovers")

(in-package :shop3-rovers)

; Load a domain and a problem defn.:
(load (asdf:system-relative-pathname "shop3" "examples/rovers/strips/domain.lisp"))
(load (asdf:system-relative-pathname "shop3" "examples/rovers/strips/p01.lisp"))
(pprint (shorter-plan (first (find-plans 'ROVERPROB1234 :verbose 0 :domain 'ROVER))))
```

**Reply by ko56** [reply]
Ah, thanks, I see.  Yes, I have checked out SHOP3 from Git, I don't have Quicklisp.

I have some more questions, if you don't mind:

1) Apparently, there a difference between (in-package :SHOP-USER) and the two lines I had, i.e.
(use-package :shop3)
(use-package :common-lisp)
So the 2nd line above puts me in 'common-lisp'.  (in-package :SHOP-USER) would not do that?
The reason I wrote those two lines is that I wanted something more transparent than (in-package :SHOP-USER), and I
thought the two 'use-package' lines would be equivalent. Is there a more transparent equivalent?

2) I have in ~/.config/common-lisp/source-registry.conf.d a file 'ko-asdf.conf' with this content:

; Look for .asd files here:
(:tree "~/build/shop3/")

Can I put some version of the (asdf:initialize-source-registry ... stuff you wrote in there?

3) Is there a way to avoid loading the "shop3/rovers" system?  And the (in-package :shop3-rovers) line?
Can one not run the planner by just supplying a domain file and a problem file?  (I've used a number of PDDL planners that work this way, so I assumed SHOP3 would work similarly.)

Thanks very much.

**Comment by rpgoldman** [comment]
Briefly (because this would require a full CL tutorial):

* `use-package` is not the same as `in-package`.  `use-package` is similar to `import` in python.  So what you have done with these two lines is to import the external symbols from the `shop3` package and `common-lisp` package into the current package (which is `common-lisp-user`). Note that importing `common-lisp` is a no-op, since SBCL's `common-lisp-user` *already* imports `common-lisp`.  So the second line *does not* put you in common-lisp.
*  The stuff I used to configure ASDF could be put in your `ko-asdf.conf`.
* Loading the rovers system ensures that the `shop3-rovers` package is created and also loads the domain definition. There isn't any reason not to do this.
* "Can one not run the planner by just supplying a domain file and a problem file? (I've used a number of PDDL planners that work this way, so I assumed SHOP3 would work similarly.)".  -- Not at present.  You interact with SHOP3 through the Lisp Read Eval Print Loop (REPL). 
  It would be relatively easy to make a command-line executable that would take a domain file and a problem file as arguments, perhaps using ASDF's build executable capabilities, but I have never done that.  I'd accept a merge request for this, but don't have any immediate plans to make one myself.  Such an executable  would probably require translating and formatting the plan(s) produced: SHOP3's `find-plans` is a *function* that *returns* plans; it is not a *program* that *prints* plans, unlike most PDDL planners.

Probably the best thing for you to do will be to have a look at [Practical Common Lisp](http://www.gigamonkeys.com/book/) and/or [Paradigms of AI Programming](https://github.com/norvig/paip-lisp) to learn your way around Common Lisp.
